File: test/baselineParity.test.py
#!/usr/bin/env python3
"""
Champion baseline parity — Python side.

Twin of `test/baselineParity.test.js`. Both assert the same
`test/fixtures/baseline_golden.json`, so the JavaScript baseline that serves the
live radar and the Python baseline that feeds the ML challenger's
`baseline_linear` feature cannot drift apart without failing CI.

Imports only `baselineScore`, never `mlChallenger`, so this runs on a bare
Python with no numpy / lightgbm / xgboost installed.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from baselineScore import (
    BASELINE_SIGMOID_SLOPE,
    BASELINE_WEIGHTS,
    baseline_columns,
    baseline_linear,
    baseline_score01,
)

GOLDEN_PATH = ROOT / "test" / "fixtures" / "baseline_golden.json"


def main() -> None:
    golden = json.loads(GOLDEN_PATH.read_text(encoding="utf-8"))
    tolerance = golden["tolerance"]

    # The fixture pins the constants too. Without this, changing a weight in
    # both the code and the fixture would pass while silently redefining the
    # champion the challenger is measured against.
    assert BASELINE_WEIGHTS == golden["weights"], (
        f"baselineScore.py weights {BASELINE_WEIGHTS} drifted from {GOLDEN_PATH.name} "
        f"{golden['weights']}"
    )
    assert BASELINE_SIGMOID_SLOPE == golden["sigmoidSlope"], (
        f"baselineScore.py sigmoid slope {BASELINE_SIGMOID_SLOPE} drifted from "
        f"{golden['sigmoidSlope']}"
    )

    checked = 0
    worst = 0.0
    for case in golden["cases"]:
        columns = baseline_columns(case["rows"])
        for index, row in enumerate(case["rows"]):
            expected = case["expected"][index]
            linear = baseline_linear(row, columns)
            score01 = baseline_score01(linear)

            for label, actual, want in (
                ("linear", linear, expected["linear"]),
                ("score01", score01, expected["score01"]),
            ):
                diff = abs(actual - want)
                worst = max(worst, diff)
                assert diff <= tolerance, (
                    f"{case['name']}[{index}] {label}: got {actual!r}, "
                    f"expected {want!r} (diff {diff} > {tolerance}). "
                    "The Python and JavaScript baselines have drifted."
                )
            checked += 1

    print(
        f"baseline parity (py) passed: {checked} rows across {len(golden['cases'])} "
        f"golden cases, worst deviation {worst:.3e} (tolerance {tolerance:.0e})"
    )


if __name__ == "__main__":
    main()

File: test/baselineParity.test.js
/**
 * Champion baseline parity — JavaScript side.
 *
 * The baseline formula is implemented twice, once per language: `sectorRadar.js`
 * serves the live radar, `baselineScore.py` feeds the ML challenger's
 * `baseline_linear` feature. Nothing in the runtime compares them, so a weight
 * changed on one side would silently make the challenger's benchmark a
 * different model from the one actually being served.
 *
 * This test and its Python twin (`test/baselineParity.test.py`) both assert
 * against `test/fixtures/baseline_golden.json`. Editing one implementation
 * without the other fails CI here.
 */

const assert = require('assert');
const fs = require('fs');
const path = require('path');
const {
  BASELINE_WEIGHTS,
  BASELINE_SIGMOID_SLOPE,
  baselineColumns,
  baselineLinear,
  baselineScore01,
  rankSectors,
} = require('../sectorRadar');

const golden = JSON.parse(fs.readFileSync(path.join(__dirname, 'fixtures', 'baseline_golden.json'), 'utf8'));
const TOLERANCE = golden.tolerance;

function close(actual, expected, message) {
  assert.ok(
    Number.isFinite(actual),
    `${message}: expected a finite number, got ${actual}`,
  );
  assert.ok(
    Math.abs(actual - expected) <= TOLERANCE,
    `${message}: got ${actual}, expected ${expected} (diff ${Math.abs(actual - expected)} > ${TOLERANCE})`,
  );
}

// The fixture pins the constants too. Without this, changing a weight in both
// the code and the fixture would pass while silently redefining the champion.
assert.deepStrictEqual(
  BASELINE_WEIGHTS,
  golden.weights,
  'sectorRadar.js weights drifted from test/fixtures/baseline_golden.json',
);
assert.strictEqual(
  BASELINE_SIGMOID_SLOPE,
  golden.sigmoidSlope,
  'sectorRadar.js sigmoid slope drifted from the golden fixture',
);

for (const testCase of golden.cases) {
  const columns = baselineColumns(testCase.rows);
  testCase.rows.forEach((row, index) => {
    const linear = baselineLinear(row, columns);
    const expected = testCase.expected[index];
    close(linear, expected.linear, `${testCase.name}[${index}] linear`);
    close(baselineScore01(linear), expected.score01, `${testCase.name}[${index}] score01`);
  });
}

// The radar's own path maps display names onto the canonical feature keys. A
// typo in that mapping would not show up above, so drive rankSectors directly
// and check it lands on the same numbers.
const sixSector = golden.cases.find(c => c.name === 'typical-six-sector');
const ranked = rankSectors(sixSector.rows.map(row => ({
  sector: row.sector,
  momentum5: row.momentum5,
  momentum20: row.momentum20,
  volumeRatio: row.volume_ratio,
  breadthAboveMA20: row.breadth_ma20,
  volatility20: row.volatility20,
})));

assert.strictEqual(ranked.length, sixSector.rows.length, 'rankSectors dropped rows');
for (const scored of ranked) {
  const index = sixSector.rows.findIndex(row => row.sector === scored.sector);
  assert.ok(index >= 0, `rankSectors returned unknown sector ${scored.sector}`);
  const expectedScore = Number((100 * sixSector.expected[index].score01).toFixed(1));
  assert.strictEqual(
    scored.score,
    expectedScore,
    `rankSectors ${scored.sector}: got ${scored.score}, expected ${expectedScore}`,
  );
}

// Ranking must be by descending score, and the display components must remain
// under the legacy names already written into data/shadow/*.jsonl.
for (let i = 1; i < ranked.length; i += 1) {
  assert.ok(ranked[i - 1].score >= ranked[i].score, 'rankSectors is not sorted by descending score');
  assert.strictEqual(ranked[i].rank, i + 1, 'rankSectors rank is not 1-based contiguous');
}
assert.deepStrictEqual(
  Object.keys(ranked[0].components).sort(),
  ['breadth', 'momentum20', 'momentum5', 'volatility', 'volumeRatio'],
  'emitted component names changed; the shadow ledgers already use the old ones',
);

console.log(`baseline parity (js) passed: ${golden.cases.length} golden cases + rankSectors end-to-end`);
